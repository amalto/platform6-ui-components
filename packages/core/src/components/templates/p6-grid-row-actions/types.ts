import { IconName, IconPrefix } from '@fortawesome/fontawesome-svg-core';
import { EventEmitter } from '@stencil/core';
import { Mode } from '../../../shared/types';
import { RowId } from '../../organisms/p6-grid/core/entities';

export interface ActionsOptions {
  create: { disabled: boolean; canUse: boolean };
  view: { disabled: boolean; canUse: boolean };
  edit: { disabled: boolean; canUse: boolean };
  rename: { disabled: boolean; canUse: boolean };
  delete: { disabled: boolean; canUse: boolean };
  export: { disabled: boolean; canUse: boolean };
  [action: string]: { disabled: boolean; canUse: boolean };
}

export interface Action {
  id: keyof ActionsOptions;
  title: string;
  icon: IconName;
  iconPrefix?: IconPrefix;
  mode?: Mode;
  emitter: () => EventEmitter<{ rowId: RowId }> | undefined;
}
