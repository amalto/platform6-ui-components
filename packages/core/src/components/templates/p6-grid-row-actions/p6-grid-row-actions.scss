:host {
  display: flex;
  flex-direction: row;
  justify-content: space-evenly;
  width: 100%;
}
