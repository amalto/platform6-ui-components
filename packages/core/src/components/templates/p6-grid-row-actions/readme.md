# p6-grid-row-actions

<!-- Auto Generated Below -->

## Properties

| Property             | Attribute       | Description                        | Type                          | Default     |
| -------------------- | --------------- | ---------------------------------- | ----------------------------- | ----------- |
| `isRowEdited`        | `is-row-edited` | Set to true if row is being edited | `boolean`                     | `false`     |
| `options`            | --              | options for the actions            | `ActionsOptions \| undefined` | `undefined` |
| `rowId` _(required)_ | `row-id`        | Id of the row                      | `string`                      | `undefined` |

## Events

| Event         | Description                                     | Type                              |
| ------------- | ----------------------------------------------- | --------------------------------- |
| `p6Cancel`    | Fires when the user clicks on the cancel action | `CustomEvent<{ rowId: string; }>` |
| `p6CloneRow`  | Fires when the user clicks on the clone action  | `CustomEvent<{ rowId: string; }>` |
| `p6DeleteRow` | Fires when the user clicks on the delete action | `CustomEvent<{ rowId: string; }>` |
| `p6EditRow`   | Fires when the user clicks on the edit action   | `CustomEvent<{ rowId: string; }>` |
| `p6ExportRow` | Fires when the user clicks on the export action | `CustomEvent<{ rowId: string; }>` |
| `p6RenameRow` | Fires when the user clicks on the rename action | `CustomEvent<{ rowId: string; }>` |
| `p6ViewRow`   | Fires when the user clicks on the view action   | `CustomEvent<{ rowId: string; }>` |

## Dependencies

### Depends on

- [p6-action](../../atoms/p6-action)
- [p6-icon](../../atoms/p6-icon)

### Graph

```mermaid
graph TD;
  p6-grid-row-actions --> p6-action
  p6-grid-row-actions --> p6-icon
  style p6-grid-row-actions fill:#f9f,stroke:#333,stroke-width:4px
```

---

_Built with [StencilJS](https://stenciljs.com/)_
