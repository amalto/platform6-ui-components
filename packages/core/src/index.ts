export { Components, JSX } from './components';
export { P6ButtonType } from './components/atoms/p6-button/p6-button';
export { P6CalendarType } from './components/atoms/p6-calendar/p6-calendar';
export { P6CheckboxValue } from './components/atoms/p6-checkbox/p6-checkbox';
export { P6InputValue } from './components/atoms/p6-input/p6-input';
export { P6InputType } from './components/atoms/p6-input/types';
export { Target } from './components/atoms/p6-link/p6-link';
export { P6RadioValue } from './components/atoms/p6-radio/p6-radio';
export { P6SelectControl, P6SelectValue } from './components/atoms/p6-select-native/p6-select-native';
export { P6SwitchValue } from './components/atoms/p6-switch/p6-switch';
export { P6TextareaValue } from './components/atoms/p6-textarea/p6-textarea';
export { P6LanguageValue } from './components/molecules/p6-language/p6-language';
export { P6TranslationControl, P6TranslationValue } from './components/molecules/p6-translation/p6-translation';
export {
  AlignColumnDetail,
  CellValueChangedDetail,
  Column,
  ColumnDefinition,
  ColumnId,
  DataItem,
  EditingCellDetail,
  FilterRowsDetail,
  MoveColumnDetail,
  ResetDefinitionsDetail,
  ResizeColumnDetail,
  Row,
  RowId,
  ShowColumnDetail,
  ShowOptionsDetail,
  SortColumnDetail,
} from './components/organisms/p6-grid/core/entities';
export { ServiceStatus } from './components/templates/p6-service-panel/service-status.enum';
export { P6Control } from './shared/form/control';
export { InvalidEventDetail, isCustomEvent, isInvalidEvent, isValidEvent, ValidEventDetail } from './shared/form/event';
export { Alignment, Mode, modes, Position, positions, Size, sizes, SortOrder } from './shared/types';
export { EnumArrayEntry, enumArrayToObject, enumToArray, EnumType } from './utils/enum';
export { LanguageCode } from './utils/language';
