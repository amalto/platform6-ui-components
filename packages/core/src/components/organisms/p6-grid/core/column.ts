import { Alignment, Direction, SortOrder } from '../../../../shared/types';
import { throwError } from '../../../../utils/error';
import { Column, ColumnDefinition, ColumnId, DataItem } from './entities';

export const MIN_WIDTH = 75;
export const DEFAULT_WIDTH = MIN_WIDTH;
export const INC_WIDTH = 10;

export function defaultGetValue<T extends DataItem>(data: T, context: ColumnDefinition<T>): string {
  if (context.field === undefined || data[context.field] === undefined) {
    return throwError(`field is undefined or doesn't exist, you should provide a 'getValue' function for ${JSON.stringify(context)}`);
  }

  return `${data[context.field as keyof T]}`;
}

export function defaultSetValue<T extends DataItem>(value: string, data: T, context: ColumnDefinition<T>): T {
  if (context.field === undefined || data[context.field] === undefined) {
    return throwError(`field is undefined or doesn't exist, you should provide a 'setValue' function for ${JSON.stringify(context)}`);
  }

  return {
    ...data,
    [context.field as keyof T]: value,
  };
}

let columnIds = 0;
export function fromDefinition<T extends DataItem>(column: ColumnDefinition<T>): Column<T> {
  // eslint-disable-next-line no-plusplus
  const id = column.id || column.field || columnIds++;

  return {
    ...column,
    id: `col-${id}`,
    editable: column.editable || false,
    hidden: column.hidden || false,
    width: column.width || 100,
    align: column.align || Alignment.center,
    sortOrder: column.sortOrder || SortOrder.none,
    disableHeaderMenu: column.disableHeaderMenu || false,
    getValue: column.getValue || defaultGetValue,
    setValue: column.setValue || defaultSetValue,
  };
}

export function isHidden<T extends DataItem>(column: Column<T>): boolean {
  return !!column.hidden;
}

export function replaceColumn<T extends DataItem>(updatedColumn: Column<T>, allColumns: Column<T>[]): Column<T>[] {
  return allColumns.map(col => (col.id === updatedColumn.id ? updatedColumn : col));
}

export function move<T extends DataItem>(id: ColumnId, direction: Direction, columns: Column<T>[]): Column<T>[] {
  const from = columns.findIndex(col => col.id === id);
  const to = from + (direction === Direction.left ? -1 : 1);

  if (from < 0 || to < 0 || to >= columns.length) {
    return columns;
  }

  const result = columns.slice();
  const item = result.splice(from, 1);

  return result.slice(0, to).concat(item).concat(result.slice(to, columns.length));
}
