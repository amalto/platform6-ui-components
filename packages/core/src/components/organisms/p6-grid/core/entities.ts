import { EventEmitter, JSX } from '@stencil/core';
import { Alignment, Direction, SortOrder } from '../../../../shared/types';

export type ColumnId = string;

export type DataItem = Record<string, unknown>;

export interface ColumnDefinition<T extends DataItem> {
  id?: ColumnId;
  label: string;
  field?: string | keyof T;
  editable?: boolean;
  filtreable?: boolean;
  sortable?: boolean;
  hidden?: boolean;
  width?: number;
  align?: Alignment;
  color?: string;
  sortOrder?: SortOrder;
  disableHeaderMenu?: boolean;
  getValue?: (data: T, context: ColumnDefinition<T>) => string;
  setValue?: (value: string, data: T, context: ColumnDefinition<T>) => T;
  cellEditor?: (data: T, context: ColumnDefinition<T>) => JSX.Element & CellEditor<T>;
  cellRenderer?: (data: T, rowId: RowId, context: ColumnDefinition<T>) => string | HTMLElement;
}

export interface Column<T extends DataItem> extends ColumnDefinition<T> {
  id: ColumnId;
  editable: boolean;
  hidden: boolean;
  width: number;
  align: Alignment;
  sortOrder: SortOrder;
  disableHeaderMenu: boolean;
  getValue: (data: T, context: ColumnDefinition<T>) => string;
  setValue: (value: string, data: T, context: ColumnDefinition<T>) => T;
}

export interface CellEditor<T extends DataItem> {
  column: Column<T>;
  row: Row<T>;
  value: string;

  p6GridCellValueChanged: EventEmitter<CellValueChangedDetail<T>>;
  p6GridEditingCell: EventEmitter<EditingCellDetail<T>>;
}
export type RowId = string;

export interface Row<T extends DataItem> {
  id: RowId;
  data: T;
}

interface ColumnEventDetail<T extends DataItem> {
  column: Column<T>;
}

export interface MoveColumnDetail<T extends DataItem> extends ColumnEventDetail<T> {
  direction: Direction;
}
export type ResizeColumnDetail<T extends DataItem> = ColumnEventDetail<T>;

export type AlignColumnDetail<T extends DataItem> = ColumnEventDetail<T>;

export type SortColumnDetail<T extends DataItem> = ColumnEventDetail<T>;

export type ShowColumnDetail<T extends DataItem> = ColumnEventDetail<T>;

export interface CellValueChangedDetail<T extends DataItem> {
  column: Column<T>;
  row: Row<T>;
  oldValue: string;
}

export enum EditingCellStatus {
  Start,
  Stop,
  Cancel,
}

export interface EditingCellDetail<T extends DataItem> {
  column: Column<T>;
  row: Row<T>;
  status: EditingCellStatus;
}

export interface ResetDefinitionsDetail {
  reset: boolean;
}

export interface ShowOptionsDetail {
  visible: boolean;
}

export interface FilterRowsDetail {
  value: string;
}
