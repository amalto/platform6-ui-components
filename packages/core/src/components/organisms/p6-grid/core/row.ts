import { SortOrder } from '../../../../shared/types';
import { Column, DataItem, Row, RowId } from './entities';

let rowIds = 0;
export function fromData<T extends DataItem>(data: T): Row<T> {
  rowIds += 1;

  return {
    id: `row-${rowIds}`,
    data,
  };
}

export function compareRow<T extends DataItem>(column: Column<T> | undefined): (a: Row<T>, b: Row<T>) => number {
  const sortOrder = column?.sortOrder || SortOrder.none;
  const factor = sortOrder === SortOrder.desc ? -1 : 1;

  return (a: Row<T>, b: Row<T>): number => {
    if (column === undefined || column.sortable === false || sortOrder === SortOrder.none) {
      return 0;
    }

    const valueA = column.getValue(a.data, column);
    const valueB = column.getValue(b.data, column);

    return valueA.localeCompare(valueB) * factor;
  };
}

export function filterBy<T extends DataItem>(searchValue: string, columns: Column<T>[]): (row: Row<T>) => boolean {
  return row => columns.filter(col => col.filtreable !== false).some(col => col.getValue(row.data, col).includes(searchValue));
}

export function rangeSelectRow<T extends DataItem>(currentRowIdSelect: RowId, selectedRows: Set<RowId>, displayedRow: Row<T>[]): Set<RowId> {
  const currentSelectedIndex = displayedRow.findIndex(row => row.id === currentRowIdSelect);

  if (currentSelectedIndex === -1) {
    return selectedRows;
  }

  const previousRowIdSelected = Array.from(selectedRows)
    .reverse()
    .find(rowId => displayedRow.findIndex(row => row.id === rowId) > -1);

  if (previousRowIdSelected === undefined) {
    return new Set(selectedRows.add(currentRowIdSelect));
  }

  const previousSelectedIndex = displayedRow.findIndex(row => row.id === previousRowIdSelected);

  let [from, to] = [previousSelectedIndex, currentSelectedIndex];
  let excludedFromSlice = currentRowIdSelect;
  if (currentSelectedIndex < previousSelectedIndex) {
    [from, to] = [to, from];
    excludedFromSlice = previousRowIdSelected;
  }

  return displayedRow
    .slice(from, to)
    .reduce((selected, cur) => {
      return cur.id === undefined ? selected : selected.add(cur.id);
    }, new Set(selectedRows))
    .add(excludedFromSlice);
}

export function replaceRow<T extends DataItem>(updatedRow: Row<T>, allRows: Row<T>[]): Row<T>[] {
  return allRows.map(row => (row.id === updatedRow.id ? updatedRow : row));
}
