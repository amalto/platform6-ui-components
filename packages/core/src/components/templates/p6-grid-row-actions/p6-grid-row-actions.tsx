import { Component, Event, EventEmitter, h, Host, JSX, Prop } from '@stencil/core';
import { Mode } from '../../../shared/types';
import { RowId } from '../../organisms/p6-grid/core/entities';
import { Action, ActionsOptions } from './types';

function onActionClick(action: Action, rowId: RowId): (event: Event) => void {
  return () => {
    action.emitter()?.emit({ rowId });
  };
}
function renderAction(action: Action, rowId: RowId, option: { disabled: boolean }): JSX.Element | null {
  return (
    <p6-action key={action.id} title={action.title} mode={action.mode} disabled={option.disabled} onClick={onActionClick(action, rowId)}>
      <p6-icon name={action.icon} iconPrefix={action.iconPrefix} />
    </p6-action>
  );
}

@Component({
  tag: 'p6-grid-row-actions',
  styleUrl: 'p6-grid-row-actions.scss',
  shadow: true,
})
export class P6GridRowActions {
  /**
   * Id of the row
   */
  @Prop() rowId!: RowId;

  /**
   * options for the actions
   */
  @Prop() options?: ActionsOptions;

  /**
   * Set to true if row is being edited
   */
  @Prop() isRowEdited = false;

  /**
   * Fires when the user clicks on the view action
   */
  @Event({ eventName: 'p6ViewRow' }) viewEmitter?: EventEmitter<{ rowId: RowId }>;

  /**
   * Fires when the user clicks on the edit action
   */
  @Event({ eventName: 'p6EditRow' }) editEmitter?: EventEmitter<{ rowId: RowId }>;

  /**
   * Fires when the user clicks on the rename action
   */
  @Event({ eventName: 'p6RenameRow' }) renameEmitter?: EventEmitter<{ rowId: RowId }>;

  /**
   * Fires when the user clicks on the clone action
   */
  @Event({ eventName: 'p6CloneRow' }) cloneEmitter?: EventEmitter<{ rowId: RowId }>;

  /**
   * Fires when the user clicks on the export action
   */
  @Event({ eventName: 'p6ExportRow' }) exportEmitter?: EventEmitter<{ rowId: RowId }>;

  /**
   * Fires when the user clicks on the delete action
   */
  @Event({ eventName: 'p6DeleteRow' }) deleteEmitter?: EventEmitter<{ rowId: RowId }>;

  /**
   * Fires when the user clicks on the cancel action
   */
  @Event({ eventName: 'p6Cancel' }) cancelEmitter?: EventEmitter<{ rowId: RowId }>;

  private cancelAction: Action = {
    id: 'cancel',
    title: 'cancel',
    icon: 'undo-alt',
    emitter: () => this.cancelEmitter,
  };

  private actions: Action[] = [
    { id: 'create', title: 'view', icon: 'search-plus', emitter: () => this.viewEmitter },
    { id: 'view', title: 'edit', icon: 'pencil-alt', mode: Mode.warning, emitter: () => this.editEmitter },
    { id: 'edit', title: 'rename', icon: 'i-cursor', mode: Mode.warning, emitter: () => this.renameEmitter },
    { id: 'rename', title: 'copy', icon: 'copy', mode: Mode.primary, iconPrefix: 'far', emitter: () => this.cloneEmitter },
    { id: 'delete', title: 'export', icon: 'download', mode: Mode.info, emitter: () => this.exportEmitter },
    { id: 'export', title: 'delete', icon: 'trash-alt', mode: Mode.danger, emitter: () => this.deleteEmitter },
  ];

  private defaultOption = { disabled: false, canUse: true };

  render(): JSX.Element {
    return (
      <Host>
        <slot>
          {this.isRowEdited
            ? renderAction(this.cancelAction, this.rowId, { disabled: false })
            : this.actions.filter(action => this.getOptionOf(action.id).canUse).map(action => renderAction(action, this.rowId, this.getOptionOf(action.id)))}
        </slot>
      </Host>
    );
  }

  private getOptionOf(action: keyof ActionsOptions): { disabled: boolean; canUse: boolean } {
    if (this.options === undefined) {
      return this.defaultOption;
    }

    return this.options[action] || this.defaultOption;
  }
}
