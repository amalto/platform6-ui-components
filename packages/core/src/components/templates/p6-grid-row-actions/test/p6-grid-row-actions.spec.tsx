import { h } from '@stencil/core';
import { newSpecPage } from '@stencil/core/testing';
import { P6GridRowActions } from '../p6-grid-row-actions';

describe('p6-grid-row-actions', () => {
  it('renders', async () => {
    const page = await newSpecPage({
      components: [P6GridRowActions],
      template: () => <p6-grid-row-actions rowId="one" />,
    });
    expect(page.root).toMatchSnapshot();
  });
});
