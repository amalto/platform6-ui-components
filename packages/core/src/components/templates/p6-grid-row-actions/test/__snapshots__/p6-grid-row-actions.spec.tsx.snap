// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`p6-grid-row-actions renders 1`] = `
<p6-grid-row-actions>
  <mock:shadow-root>
    <slot>
      <p6-action title="view">
        <p6-icon name="search-plus"></p6-icon>
      </p6-action>
      <p6-action mode="1" title="edit">
        <p6-icon name="pencil-alt"></p6-icon>
      </p6-action>
      <p6-action mode="1" title="rename">
        <p6-icon name="i-cursor"></p6-icon>
      </p6-action>
      <p6-action mode="5" title="copy">
        <p6-icon iconprefix="far" name="copy"></p6-icon>
      </p6-action>
      <p6-action mode="3" title="export">
        <p6-icon name="download"></p6-icon>
      </p6-action>
      <p6-action mode="0" title="delete">
        <p6-icon name="trash-alt"></p6-icon>
      </p6-action>
    </slot>
  </mock:shadow-root>
</p6-grid-row-actions>
`;
